### PR TITLE
config: sync users with config on reload

### LINKED ISSUE #11827: config: users are not synced with config on reload
If a user is removed from `credentials.users` in config and the instance is reloaded, the user still exists in `box`.
This leaves stale users in the system, which is undesirable from a security perspective.

### Example

Initial config:

```yaml
credentials:
  users:
    alice:
      password: ALICE
    bob:
      password: BOB
```

After applying, both `alice` and `bob` exist:

```lua
box.schema.user.exists('alice') -- true
box.schema.user.exists('bob')   -- true
```

Then we update config and remove `alice`:

```yaml
credentials:
  users:
    bob:
      password: BOB
```

Expected behavior after reload:

```lua
box.schema.user.exists('alice') -- false (user dropped)
box.schema.user.exists('bob')   -- true
```

Actual behavior:

```lua
box.schema.user.exists('alice') -- true  (stale user remains)
box.schema.user.exists('bob')   -- true
```

---

[TNTP-2155](https://jira.vk.team/browse/TNTP-2155)

### PR DESCRIPTION
**Pull request by mandesero**
Previously, if a user or role was removed from `credentials.users` or `credentials.roles` in config, it still remained in the instance after reload. This was intentional to prevent bricking when an empty `credentials` section was rolled out, but it left stale accounts and roles in the system, which is a security risk.

Now both users and roles are synchronized with config on reload:
 - users/roles missing from config are dropped,
 - users/roles present in config are created/updated,
 - manually created users/roles remain untouched.
 
 Closes #11827

### CODE REVIEW

File: src/box/lua/config/applier/credentials.lua [hunk 1]
```diff
@@ -834,6 +835,27 @@ end
...
+            log.verbose('credentials.apply: drop user %q', name)
+            box.schema.user.drop(name)
+        end
+    end
+
+    applied_users = {}
+    for name in pairs(user_map) do
+        applied_users[name] = true
+    end
+end
```
Comment (Totktonada) [inline src/box/lua/config/applier/credentials.lua]: It would be a nice solution, but sadly it doesn't work over restart: the `applied_users` is `nil` in the case, while some of the users are possibly created by the YAML config machinery. Also, it possibly works differently depending of whether we see or not see intermediate configurations.

We need to persist something about users created from the YAML configuration. We did something like that for `_priv`: it tracks an origin of the grant/revoke call. We can do the same for `_user`.

Also, the quoted logic misses an important scenario, when the same user is created from the YAML configuration and manually using `box.schema.user.create()`. We should keep it until 1) it is removed from the YAML configuration, 2) `box.schema.user.drop()` is called.

And there is yet another important scenario: if a user is created from YAML and Lua both, we possibly grant some privileges to the user from the YAML configuration. We should revoke them if the user is removed from the YAML configuration. Not sure it actually works this way.

Comment (Totktonada) [reply src/box/lua/config/applier/credentials.lua]: (I understand that it is a draft PR and possibly just an experiment, but I would like to let us be in sync sooner.)

File: src/box/lua/config/applier/credentials.lua [hunk 2]
```diff
@@ -887,6 +909,7 @@ local function sync_object(obj_to_sync)
             create_users(credentials.users)
 
             sync_privileges(credentials)
+            sync_users(credentials.users)
```
Comment (Totktonada) [inline src/box/lua/config/applier/credentials.lua]: Hm. I never said that (my bad), but it would be intuitive to do the same for roles.

File: src/box/lua/config/applier/credentials.lua [hunk 3]
```diff
@@ -834,6 +846,42 @@ end
...
+    for _, tuple in box.space._user:pairs() do
+        if tuple.type == 'user' and tuple[USER_OPTS_FIELD_ID] ~= nil and
+           tuple[USER_OPTS_FIELD_ID].origins ~= nil and
+           tuple[USER_OPTS_FIELD_ID].origins[CONFIG_ORIGIN] and
+           (user_map or {})[tuple.name] == nil then
+            table.insert(to_drop, tuple.name)
+        end
+    end
+    for _, name in ipairs(to_drop) do
+        log.info('credentials.apply: drop user %q', name)
```
Comment (mandesero) [inline src/box/lua/config/applier/credentials.lua]: Replace `log.info` with `log.verbose`

File: src/box/lua/schema.lua [hunk 4]
```diff
@@ -3684,13 +3685,65 @@ box.schema.user.passwd = function(name, new_password)
...
+    end
+    if tuple[USER_OPTS_FIELD_ID] == nil or
+       tuple[USER_OPTS_FIELD_ID].origins == nil then
+        return {[DEFAULT_ORIGIN] = true}
+    end
+    local res = {}
+    for name, _ in pairs(tuple[USER_OPTS_FIELD_ID].origins) do
+        res[name] = true
+    end
+    return res
```
Comment (Totktonada) [inline src/box/lua/schema.lua]: Nit: It looks like an equivalent to just `return tuple[USER_OPTS_FIELD_ID].origins`. If there is a reason to write it with a loop, please, leave a remark about it.

File: src/box/lua/schema.lua [hunk 5]
```diff
@@ -3684,13 +3685,65 @@ box.schema.user.passwd = function(name, new_password)
...
+end
+
+-- Return opts field for user/role based on origins map.
+local function user_opts_from_origins(origins)
+    local normalized = {}
+    for name, v in pairs(origins) do
+        if v then
+            normalized[name] = true
+        end
+    end
```
Comment (Totktonada) [inline src/box/lua/schema.lua]: Nit: `v` can't be `nil` in the if-condition, because a table can't consist a `nil` value. It seems that the `normalized` table is always an equivalent to the input table `origins`.

Comment (Totktonada) [reply src/box/lua/schema.lua]: It seems, this code aims to handle `false` value. I'm in doubt that we should use/support it. However, if we really need, let's leave some remark here.

File: src/box/lua/schema.lua [hunk 6]
```diff
@@ -3684,13 +3685,65 @@ box.schema.user.passwd = function(name, new_password)
...
+    if origins[origin] then
+        return
+    end
+    origins[origin] = true
+    local opts_field = user_opts_from_origins(origins)
+    if opts_field ~= nil then
+        _user:update(uid, {{'=', USER_OPTS_FIELD_ID, opts_field}})
+    else
+        if tuple[USER_OPTS_FIELD_ID] ~= nil then
+            _user:update(uid, {{'#', USER_OPTS_FIELD_ID, 1}})
```
Comment (Totktonada) [inline src/box/lua/schema.lua]: Thinking about extendability (a kind of forward compatibility), I would highlight that this operation moves fields beyond `opts` one position left. It seems it is unintended effect.

File: src/box/lua/schema.lua (hunk 4 above)
Comment (mandesero) [reply src/box/lua/schema.lua]: Agreed, simplified it

File: src/box/lua/schema.lua (hunk 5 above)
Comment (mandesero) [reply src/box/lua/schema.lua]: I agree, we don't expect false values ​​in the map. Fixed.

File: src/box/lua/schema.lua (hunk 6 above)
Comment (mandesero) [reply src/box/lua/schema.lua]: I updated the code to keep the field position constant and just clear its value:
```lua
_user:update(uid, {{'=', USER_OPTS_FIELD_ID, box.NULL}})
```

File: src/box/lua/schema.lua [hunk 7]
```diff
@@ -3684,13 +3685,55 @@ box.schema.user.passwd = function(name, new_password)
...
-    local template = {password = 'string', if_not_exists = 'boolean'}
+    local template = {password = 'string', if_not_exists = 'boolean',
+                      _origin = 'string'}
     check_param_table(opts, template, 2)
+    local origin = opts._origin or DEFAULT_ORIGIN
     if uid then
+        user_add_origin(uid, origin)
         if not opts.if_not_exists then
             box.error(box.error.USER_EXISTS, name, 2)
         end
```
Comment (Totktonada) [inline src/box/lua/schema.lua]: I think that origins should work like [tenants](https://www.cloudflare.com/learning/cloud/what-is-multitenancy/) in regards to the `if_not_exists` and `if_exists` options (their `false` values to be precise).

IOW, `create(<...>, {_origin = {foo = true}})` only reports an error if the user exists in the `foo` origin. It doesn't matter if it exists in other origins. Without `opts._origins` it should check only `''` (default) origin.

Similarly for `drop(<...>, {_origin = {foo = true}})`. Should check existence only in the `foo` origin.

----

I guess you don't need to add `if_not_exists` and `if_exists` in `applier/credentials.lua` if the `box/lua/schema.lua` code works this way.

File: changelogs/unreleased/gh-11827-sync-users-after-reload.md
Comment (Totktonada) [inline changelogs/unreleased/gh-11827-sync-users-after-reload.md:1]: I would add a highlight into the release notes[^1] that describes recommended manual actions after installing this update. I think that we should recommend to check that no 'orphan' user/role entries are present. We can provide a script that lists all users and roles that are not tracked by config's machinery (for example, put it to the [wiki](https://github.com/tarantool/tarantool/wiki)). The script can be linked from the release notes.

It is especially important, because users/roles created by previous versions of tarantool are not marked as tracked by config's machinery. So, they are not deleted even if they're removed from the configuration.

[^1]: After the `----` marker, see the syntax documentation [here](https://github.com/tarantool/tarantool/blob/4a329cea763459ff63634c0bf3fde326e4931685/doc/changelogs.md#leave-a-note-to-a-release-manager).

File: src/box/lua/config/applier/credentials.lua [hunk 8]
```diff
@@ -950,6 +1000,10 @@ local triggers_are_set
 local function apply(config_module)
     config = config_module
 
+    if config._configdata:get('credentials') == nil then
+        return true
+    end
```
Comment (Totktonada) [inline src/box/lua/config/applier/credentials.lua]: What is the idea behind this condition? I think it worth to have a remark.

Is it to support the original idea about 'an empty config shouldn't drop all the accesses'? I would say that a motivation behind it is vague. As for me, it is not enough to have this exceptional case.

First, it was originally discussed in context of an *entirely empty* config, based on moonlibs/config experience. However, tarantool 3.x config can't be entirely empty. In this case a tarantool instance can't find itself in the configuration and refuses to start. The database files remain untouched.

Second, we're thinking about a misconfiguration and want to protect a user from the service outage. If we throw out the users/roles, that's the outage, right. The same about data directories. And memory parameters (if the database is large enough). And likely `app`/`roles`/`roles_cfg` sections. I don't think that the `credentials` section is specific in this regard. If we want to protect a user from such kind of mistakes, we need to think about it on some more general level.

File: src/box/lua/schema.lua (hunk 7 above)
Comment (mandesero) [reply src/box/lua/schema.lua]: ~~Kept `if_not_exists = true` here~~
https://github.com/tarantool/tarantool/blob/5a52d71841ecc7ae36656530e30237d4064da307/src/box/lua/config/applier/credentials.lua#L733-L747
~~so that if a user/role already exists manually, the applier still attaches `origin = 'config'`. Without this, such objects wouldn’t be tracked by config.~~

Comment (mandesero) [reply src/box/lua/schema.lua]: ~~Also I updated `user/role.create` so that `if_not_exists` is checked per origin rather than globally.~~

* ~~If the role/user already has the requested origin then respect `if_not_exists` and raise if it’s false.~~
* ~~If the role/user exists but **without** this origin then we extend its origins instead of erroring, so that config/manual origins are tracked correctly.~~

File: src/box/lua/config/applier/credentials.lua (hunk 8 above)
Comment (mandesero) [reply src/box/lua/config/applier/credentials.lua]: After personal discussions with @Totktonada I clarified the real problem with starting an instance when `credentials` is empty: the applier tries to drop system users/roles and fails.

The proposed solution is to normalize `credentials` in this case by filling in default system users and roles, so they are never treated as "removed from config".

The old logic _"do nothing if credentials is empty"_ was removed.

File: changelogs/unreleased/gh-11827-sync-users-after-reload.md
Comment (mandesero) [reply changelogs/unreleased/gh-11827-sync-users-after-reload.md:1]: Great idea, I added an example script to `doc/examples/config/list-origins.lua`.
It lists all non-system users and roles with their origins and prints drop commands for those not tracked by the config.

Here is a sample output:

```
tarantool> dofile("doc/examples/config/list-origins.lua")
All non-system users:
  bob (origins: config)
  alice (origins: manual, config)
  manual-user (origins: manual)

All non-system roles:
  config-role-1 (origins: config)
  manual-role (origins: manual)

Drop commands for non-config origins (users):
  box.schema.user.drop("alice")
  box.schema.user.drop("manual-user")

Drop commands for non-config origins (roles):
  box.schema.role.drop("manual-role")
```

This is still an experimental version -- we need to think about the most convenient output format and prepare a proper wiki page.

File: test/config-luatest/gh_10182_roles_before_box_cfg_test.lua [hunk 9]
```diff
@@ -112,7 +112,7 @@ tg.test_role_with_early_load_tag_added = function(g)
     t.assert(g.server:grep_log(
         'Role "two" with the "early_load" tag was added to the config, it ' ..
         'cannot be loaded before the first box.cfg call',
-        1024, {filename = g.server.chdir .. '/tarantool.log'}))
+        2048, {filename = g.server.chdir .. '/tarantool.log'}))
```
Comment (mandesero) [inline test/config-luatest/gh_10182_roles_before_box_cfg_test.lua]: I bump the read window, because after we added the new warning, the target error line no longer fits into the last 1KB tail, so `grep log` missed it. Similar for `config_include_test.lua`.

File: src/box/lua/config/applier/credentials.lua [hunk 10]
```diff
@@ -834,6 +840,92 @@ end
...
+            -- or fetch the latest version:
+            -- %s
+        ]]):format(script_url)) .. '\n'
+
+        config._aboard:set({
+            type = 'warn',
+            message = msg,
+        }, {key = orphan_users_alert_key})
+
+        config._aboard:drop(orphan_users_alert_key)
```
Comment (Totktonada) [inline src/box/lua/config/applier/credentials.lua:906]: What is the point to set the warning and then immediatelly drop it? It looks like the `else` keyword is forgotten in between.

File: test/config-luatest/gh_10182_roles_before_box_cfg_test.lua (hunk 9 above)
Comment (Totktonada) [reply test/config-luatest/gh_10182_roles_before_box_cfg_test.lua]: Is the warning pops up due to #11848?

I guess it shouldn't appear on a freshly created database after it is fixed.

File: src/box/lua/schema.lua [hunk 11]
```diff
@@ -3952,7 +4016,19 @@ local function drop(uid, level)
...
+            -- _vpriv consistent.
+            for origin, _ in pairs(tuple[PRIV_OPTS_FIELD_ID].origins) do
+                revoke(level + 1, tuple.grantee, tuple.grantee, uid,
+                       nil, nil, {_origin = origin})
+            end
+        else
+            revoke(level + 1, tuple.grantee, tuple.grantee, uid,
+                   nil, nil, {_origin = DEFAULT_ORIGIN})
+        end
     end
```
Comment (Totktonada) [inline src/box/lua/schema.lua:4028]: It feels unecessarily complicated and fragile for me.

First, `box.schema.{user,role}.grant()` does not check that the user/role to whom the role is granted exists *in the given origin*. So, it is possible that we leave orphan entries in `_priv`.

Second, if this is not the case, these remove-one-by-one finally goes to `box.space._priv:delete()`. So why don't just call it directly?

The same for the privilege revoke below in this function.

File: src/box/lua/schema.lua [hunk 12]
```diff
@@ -3970,12 +4046,69 @@ local function drop(uid, level)
...
+        return
+    end
+    local _vpriv = box.space[box.schema.VPRIV_ID]
+    for _, t in pairs(_vpriv.index.primary:select{uid}) do
+        local priv = 0
+        if t[PRIV_OPTS_FIELD_ID] ~= nil
+            and t[PRIV_OPTS_FIELD_ID].origins ~= nil then
+            priv = t[PRIV_OPTS_FIELD_ID].origins[origin] or 0
+        elseif origin == DEFAULT_ORIGIN then
+            priv = t.privilege
```
Comment (Totktonada) [inline src/box/lua/schema.lua:4091]: `t.privilege` is a sum of privileges from all the origins. Using it for the default origin feels incorrect.

File: src/box/lua/schema.lua [hunk 13]
```diff
@@ -3970,12 +4046,69 @@ local function drop(uid, level)
...
+        end
+        if priv ~= 0 then
+            box.session.su('admin', revoke, level + 2, uid, uid, priv,
```
Comment (Totktonada) [inline src/box/lua/schema.lua:4096]: Nit: `level + 1` above in this function, but `level + 2` here. Should be the same?

File: src/box/lua/schema.lua [hunk 14]
```diff
@@ -3952,7 +4016,19 @@ local function drop(uid, level)
...
     for _, tuple in pairs(grants) do
-        revoke(level + 1, tuple.grantee, tuple.grantee, uid)
+        if tuple[PRIV_OPTS_FIELD_ID] ~= nil and
+           tuple[PRIV_OPTS_FIELD_ID].origins ~= nil then
+            -- A role/user may be granted both manually and via config at the
+            -- same time, so we must revoke each origin separately to keep
+            -- _vpriv consistent.
+            for origin, _ in pairs(tuple[PRIV_OPTS_FIELD_ID].origins) do
+                revoke(level + 1, tuple.grantee, tuple.grantee, uid,
+                       nil, nil, {_origin = origin})
```
Comment (Totktonada) [inline src/box/lua/schema.lua]: Nit: `level + 2` below in this function, but `level + 1` here. Should be the same (`level + 2`)?

File: src/box/lua/schema.lua (hunk 12 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:4091]: I see, it is only if there is no origin field in the privilege entry (so we have only `t.privilege`). OK so. But I would leave some comment here.

File: src/box/lua/schema.lua [hunk 15]
```diff
@@ -3970,12 +4046,69 @@ local function drop(uid, level)
...
+            priv = t[PRIV_OPTS_FIELD_ID].origins[origin] or 0
+        elseif origin == DEFAULT_ORIGIN then
+            priv = t.privilege
+        end
+        if priv ~= 0 then
+            box.session.su('admin', revoke, level + 2, uid, uid, priv,
+                           t.object_type, t.object_id,
+                           {_origin = origin, if_exists = true})
+        end
+    end
```
Comment (Totktonada) [inline src/box/lua/schema.lua:4100]: I'm still in some doubt regarding revoking privileges granted to user U in origin O, when the user U is dropped in origin O. Currently `box.schema.user.grant()` allows to grant a privilege in origin O to a user from another origin (without origin O).

However, if we're looking on these origins as for tenants, it makes sense.

I'll spin around these thoughts for a while.

File: test/config-luatest/credentials_applier_test.lua [hunk 16]
```diff
@@ -493,7 +502,9 @@ g.test_remove_user_role = function(g)
                 }
             }
         },
-        verify_2 = verify,
+        -- Legacy behavior: before gh-11828, removing a user or role
+        -- from the config did NOT delete it from box.
```
Comment (Totktonada) [inline test/config-luatest/credentials_applier_test.lua]: Nit: I read it like this legacy behavior is checked here.

File: src/box/lua/schema.lua [hunk 17]
```diff
@@ -3718,13 +3769,26 @@ box.schema.user.create = atomic_wrapper(function(name, opts)
...
+        origin = opts._origin
+    end
+    local uid = user_resolve(name, 2)
+    if uid == nil then
+        return false
+    end
+    if origin == nil then
         return true
-    else
+    end
```
Comment (Totktonada) [inline src/box/lua/schema.lua:3792]: To think: is it OK that `opts = nil` and `opts = {_origin = ''}` works differently here? It works the same for all the other functions that accept `opts = {_origin = <...>}`.

I guess that we really need that `box.schema.{user,role}.exists()` checks a real presence of a user, not just its presence in the default origin, because it may be used in some privilege check logic, for example.

Need to spin around those thoughs for a while.

File: src/box/lua/config/applier/credentials.lua (hunk 8 above)
Comment (Totktonada) [reply src/box/lua/config/applier/credentials.lua]: > The proposed solution is to normalize `credentials` in this case by filling in default system users and roles, so they are never treated as "removed from config".

Why *system* users/roles are ever considered in this logic? `has_unmanaged_subjects()` already has the `SYSTEM_USER_ID_MAX` check. Can we just skip system users/roles in the `drop_{users,roles}_not_in_config()` logic too?

I understand why we do that before the new origins logic, but now it seems unnecessary.

File: test/config-luatest/gh_10182_roles_before_box_cfg_test.lua (hunk 9 above)
Comment (mandesero) [reply test/config-luatest/gh_10182_roles_before_box_cfg_test.lua]: Yes, I checked it with #11848 and you are right.

File: src/box/lua/schema.lua [hunk 18]
```diff
@@ -3684,15 +3685,65 @@ box.schema.user.passwd = function(name, new_password)
...
+    local _user = box.space[box.schema.USER_ID]
+    local tuple = _user:get{uid}
+    local origins = user_origins_from_tuple(tuple)
+    if origins[origin] then
+        return
+    end
+    origins[origin] = true
+    local opts_field = user_opts_from_origins(origins)
+    if opts_field ~= nil then
+        _user:update(uid, {{'=', USER_OPTS_FIELD_ID, opts_field}})
```
Comment (georgiy-belyanin) [inline src/box/lua/schema.lua:3716]: It does not seem this update follows the `_user` space format. I am not sure: how it affects schema downgrade/upgrade?

Comment (georgiy-belyanin) [reply src/box/lua/schema.lua:3716]: I mean it is ok if this fields are simply ignored if not present. I am just not very familiar with handling space formats and downgrading process and I want to be sure nothing is broken due to new field.

File: changelogs/unreleased/gh-11827-sync-users-after-reload.md [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+----
+
+Manual action required: ensure that all users and roles are managed by the
+YAML configuration.
+
+A customer has to run the following script [1] to ensure that all the users
+and roles are managed solely by the YAML configuration.
+
+It allows to identify users/roles that are kept forever, because created from
+Lua (or from config on tarantool version less than 3.6.0) and decide, whether
```
Comment (maryiaLichko) [inline changelogs/unreleased/gh-11827-sync-users-after-reload.md]:
```suggestion
Lua (or from config on tarantool version less than 3.6.0), and decide whether
```

**Review by maryiaLichko — APPROVED** [review]
chagelog is ok for me. if changed -> request review again

File: test/config-luatest/credentials_applier_test.lua (hunk 16 above)
Comment (mandesero) [reply test/config-luatest/credentials_applier_test.lua]: I tried to rewrite it :) Maybe it's clearer now.

File: test/config-luatest/gh_10182_roles_before_box_cfg_test.lua (hunk 9 above)
Comment (mandesero) [reply test/config-luatest/gh_10182_roles_before_box_cfg_test.lua]: I think I can resolve this conversation since there are no more changes.

File: src/box/lua/schema.lua (hunk 18 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:3716]: We can update the `_user` format, but we don't have to do it. In my opinion, if we can skip the schema upgrade, it is easier for a user (less manual actions are needed after the executable update).

File: src/box/lua/config/applier/credentials.lua (hunk 10 above)
Comment (Totktonada) [reply src/box/lua/config/applier/credentials.lua:906]: Seems resolved now.

File: src/box/lua/schema.lua (hunk 17 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:3792]: At least now it is explicitly commented. OK.

File: src/box/lua/schema.lua (hunk 11 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:4028]: Seems solved.

File: src/box/lua/schema.lua (hunk 12 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:4091]: I see the comment now, thanks!

File: src/box/lua/schema.lua (hunk 15 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:4100]: The solution seems OK. Please, leave some remark close to this logic, what is does and why. If you mention corner cases as well, it would be nice.

File: test/config-luatest/credentials_applier_test.lua [hunk 20]
```diff
@@ -1519,22 +1533,46 @@ g.test_set_user_instead_of_role = function(g)
     cluster['i-001']:exec(function()
         box.schema.user.create('role_two')
         local info = require('config'):info()
-        local exp = 'box.schema.user.grant("user_one", "execute", "role", ' ..
+        local exp_grant =
+            'box.schema.user.grant("user_one", "execute", "role", ' ..
             '"role_two") has failed because either the object has not been ' ..
             'created yet, a database schema upgrade has not been performed, ' ..
             'or the privilege write has failed (separate alert reported)'
+        local exp_orphan_sub = 'Found users/roles authored from Lua and ' ..
```
Comment (Totktonada) [inline test/config-luatest/credentials_applier_test.lua]: Maybe adjust `box.schema.user.create` (add `_origin`) instead?

Comment (mandesero) [reply test/config-luatest/credentials_applier_test.lua]: This is a great solution, thank you!

File: src/box/lua/schema.lua (hunk 15 above)
Comment (mandesero) [reply src/box/lua/schema.lua:4100]: Done.

File: src/box/lua/schema.lua (hunk 14 above)
Comment (mandesero) [reply src/box/lua/schema.lua]: I close this thread since there's no longer `level` in `full_drop`

File: src/box/lua/schema.lua (hunk 13 above)
Comment (mandesero) [reply src/box/lua/schema.lua:4096]: `box.session.su` already adds +1 to the call stack level, and `revoke` itself adds another +1. So `level + 2` is the correct value here.

File: src/box/lua/config/applier/credentials.lua (hunk 8 above)
Comment (mandesero) [reply src/box/lua/config/applier/credentials.lua]: I updated `get_credentials`, now there are no default users except `sharding`, also `drop` methods does not touch system records

File: src/box/lua/schema.lua (hunk 13 above)
Comment (Totktonada) [reply src/box/lua/schema.lua:4096]: Forgot about `box.session.su`, agree.

File: test/config-luatest/credentials_sync_test.lua [hunk 21]
```diff
@@ -0,0 +1,203 @@
+local t = require('luatest')
+local cbuilder = require('luatest.cbuilder')
+local cluster = require('test.config-luatest.cluster')
```
Comment (Totktonada) [inline test/config-luatest/credentials_sync_test.lua]: Nit: We can use `luatest.cluster`. AFAIR, the all the differences are the following:

* No need to manually add `cluster.{init,drop,clean}` to the luatest hooks, because luatest performs it on its own.
* We have to call `cluster:new()` instead of `cluster.new()`.

In fact, it would be nice to move all the tests to `luatest.cluster` and drop the helper module from tarantool repository.

File: tools/find-orphan-users.lua [hunk 22]
```diff
@@ -0,0 +1,178 @@
+
```
Comment (Totktonada) [inline tools/find-orphan-users.lua]: Nit: Why the first line is empty?

Comment (mandesero) [reply tools/find-orphan-users.lua]: That was unintentional. Fixed.

File: test/config-luatest/credentials_sync_test.lua (hunk 21 above)
Comment (mandesero) [reply test/config-luatest/credentials_sync_test.lua]: Fixed.

Comment (Totktonada) [reply test/config-luatest/credentials_sync_test.lua]: Thanks! It seems, I can't resolve this comment now (left the organization on GitHub). Please, proceed.

File: tools/find-orphan-users.lua (hunk 22 above)
Comment (Totktonada) [reply tools/find-orphan-users.lua]: Can't resolve it, please, proceed.

File: tools/find-orphan-users.lua [hunk 23]
```diff
@@ -0,0 +1,177 @@
+-- Print commands to drop users and roles that are not managed by
```
Comment (georgiy-belyanin) [inline tools/find-orphan-users.lua:1]: It would be nice if this script was somehow tested within some of the `credential_sync_test` scenarios.

File: src/box/lua/schema.lua [hunk 24]
```diff
@@ -3967,15 +4039,73 @@ local function drop(uid, level)
...
+            -- New-style entry with per-origin privileges.
+            -- Revoke only the part that belongs to the given origin.
+            priv = t[PRIV_OPTS_FIELD_ID].origins[origin] or 0
+        elseif origin == DEFAULT_ORIGIN then
+            -- Legacy entry without origins map: treat it as belonging to the
+            -- default `DEFAULT_ORIGIN` origin and revoke the whole privilege.
+            priv = t.privilege
+        end
+        if priv ~= 0 then
+            box.session.su('admin', revoke, level + 2, uid, uid, priv,
```
Comment (georgiy-belyanin) [inline src/box/lua/schema.lua:4096]: Nit: I think it is worth to leave here the same comment on the unnecessary check and need for additional `box.session.su` here.

**Review by georgiy-belyanin — APPROVED** [review]
Thank you! This is some really neat work!
